### PR TITLE
[Merged by Bors] - feat(Data/Nat/GCD/BigOperators): add lemmas about coprimality with `List.prod`, `Multiset.prod`, `Finset.prod`

### PR DESCRIPTION
**Pull request by iehality**
Add `coprime_xxx_prod_left_iff`, `coprime_xxx_prod_right_iff` lemma for `List`, `Multiset`, `Finset` and `Fintype`. This is a PR separated from #8887.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Nat/GCD/BigOperators.lean [hunk 1]
```diff
@@ -30,4 +30,38 @@ theorem coprime_prod_right {ι : Type*} {x : ℕ} {s : ι → ℕ} {t : Finset 
   Finset.prod_induction s (fun y ↦ x.Coprime y) (fun a b ↦ Coprime.mul_right) (by simp)
 #align nat.coprime_prod_right Nat.coprime_prod_right
 
+theorem coprime_prod_left_iff {ι : Type*} [DecidableEq ι] {x : ℕ} {s : ι → ℕ} {t : Finset ι} :
+    Coprime (∏ i : ι in t, s i) x ↔ (∀ i : ι, i ∈ t → Coprime (s i) x) :=
+  Finset.induction_on t (by simp) fun i s his ih ↦ by
+    rw[Finset.prod_insert his, Nat.coprime_mul_iff_left, ih, Finset.forall_mem_insert]
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/GCD/BigOperators.lean:36]: A general style comment: spaces after `rw` and `simp`:
```suggestion
    rw [Finset.prod_insert his, Nat.coprime_mul_iff_left, ih, Finset.forall_mem_insert]
```

**Comment by tb65536** [comment]
Can you use your lemmas to shorten the proofs of the existing lemmas?

**Review by tb65536 — COMMENTED** [review]
Can you use your lemmas to shorten the proofs of the existing lemmas?

File: Mathlib/Data/Nat/GCD/BigOperators.lean [hunk 2]
```diff
@@ -30,4 +30,38 @@ theorem coprime_prod_right {ι : Type*} {x : ℕ} {s : ι → ℕ} {t : Finset 
...
+
+theorem coprime_prod_right_iff {ι : Type*} [DecidableEq ι] {x : ℕ} {s : ι → ℕ} {t : Finset ι} :
+    Coprime x (∏ i : ι in t, s i) ↔ (∀ i : ι, i ∈ t → Coprime x (s i)) :=
+  Finset.induction_on t (by simp) fun i s his ih ↦ by
+    rw[Finset.prod_insert his, Nat.coprime_mul_iff_right, ih, Finset.forall_mem_insert]
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/GCD/BigOperators.lean:41]:
```suggestion
  Finset.cons_induction_on t (by simp) fun i s his ih ↦ by
    rw [Finset.prod_cons, Nat.coprime_mul_iff_right, ih, Finset.forall_mem_cons]
```
avoids the need for `DecidableEq ι`

File: Mathlib/Data/Nat/GCD/BigOperators.lean (hunk 1 above)
Comment (iehality) [reply Mathlib/Data/Nat/GCD/BigOperators.lean]: I corrected that notation.

File: Mathlib/Data/Nat/GCD/BigOperators.lean (hunk 2 above)
Comment (iehality) [reply Mathlib/Data/Nat/GCD/BigOperators.lean]: I removed `DecidableEq`.

File: Mathlib/Data/Nat/GCD/BigOperators.lean [hunk 3]
```diff
@@ -30,4 +30,38 @@ theorem coprime_prod_right {ι : Type*} {x : ℕ} {s : ι → ℕ} {t : Finset 
   Finset.prod_induction s (fun y ↦ x.Coprime y) (fun a b ↦ Coprime.mul_right) (by simp)
 #align nat.coprime_prod_right Nat.coprime_prod_right
 
+theorem coprime_prod_left_iff {ι : Type*} {x : ℕ} {s : ι → ℕ} {t : Finset ι} :
+    Coprime (∏ i : ι in t, s i) x ↔ (∀ i : ι, i ∈ t → Coprime (s i) x) :=
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/GCD/BigOperators.lean:34]:
```suggestion
    Coprime (∏ i in t, s i) x ↔ ∀ i ∈ t, → Coprime (s i) x :=
```
and below

Comment (iehality) [reply Mathlib/Data/Nat/GCD/BigOperators.lean]: changed notation.

File: Mathlib/Data/Nat/GCD/BigOperators.lean [hunk 4]
```diff
@@ -30,4 +30,38 @@ theorem coprime_prod_right {ι : Type*} {x : ℕ} {s : ι → ℕ} {t : Finset 
   Finset.prod_induction s (fun y ↦ x.Coprime y) (fun a b ↦ Coprime.mul_right) (by simp)
```
Comment (tb65536) [inline Mathlib/Data/Nat/GCD/BigOperators.lean:30]: Could you move these two lemmas below and shorten their proofs?

Comment (eric-wieser) [reply Mathlib/Data/Nat/GCD/BigOperators.lean:30]: I think using `alias` would be appropriate here, and perhaps renaming to `Coprime.prod_left` etc.

Comment (iehality) [reply Mathlib/Data/Nat/GCD/BigOperators.lean]: I redefined using alias and renamed to `Coprime.prod_left` and `Coprime.prod_right`.

**Comment by eric-wieser** [comment]
bors d+

Thanks! I pushed some tweaks to the proofs, but this is ready to go assuming you're happy with them and CI passes.

**Comment by iehality** [comment]
I am satisfied with the changes. Thanks!

bors merge

**Pull request closed without merging**